pkg/clusteragent/admission/mutate/cws_instrumentation.go

```diff
@@ -0,0 +1,314 @@
+	if !isPodCWSInstrumentationReady(pod.GetAnnotations()) {
+		// pod isn't instrumented, do not attempt to override the pod exec command
+		log.Debugf("Ignoring exec request into %s, pod not instrumented yet", podString(pod))
+		return nil
+	}
+
+	// prepare the user session context
+	userSessionCtx, err := usersessions.PrepareK8SUserSessionContext(userInfo, cwsUserSessionDataMaxSize)
+	if err != nil {
+		metrics.MutationErrors.Inc(metrics.CWSExecInstrumentation, "cannot serialize user info", "")
+		log.Debugf("ignoring instrumentation of %s: %v", podString(pod), err)
+		return err
+	}
+
+	if len(exec.Command) > 7 {
+		// make sure the command hasn't already been instrumented
+		if exec.Command[0] == filepath.Join(cwsMountPath, "cws-instrumentation") &&
+			exec.Command[1] == "inject" &&
+			exec.Command[2] == "--session-type" &&
+			exec.Command[3] == "k8s" &&
+			exec.Command[4] == "--data" &&
+			exec.Command[6] == "--" {
+
+			if exec.Command[5] == string(userSessionCtx) {
+				log.Debugf("Exec request into %s is already instrumented, ignoring", podString(pod))
+				injected = true
+				return nil
+			}
+		}
+	}
+
+	// override the command with the call to cws-instrumentation
+	exec.Command = append([]string{
+		filepath.Join(cwsMountPath, "cws-instrumentation"),
+		"inject",
+		"--session-type",
+		"k8s",
+		"--data",
+		string(userSessionCtx),
+		"--",
+	}, exec.Command...)
+
+	log.Debugf("Pod exec request to %s is now instrumented for CWS", podString(pod))
+	injected = true
+
+	return nil
+}
+
+// InjectCWSPodInstrumentation injects CWS pod instrumentation
+func InjectCWSPodInstrumentation(rawPod []byte, _ string, ns string, _ *authenticationv1.UserInfo, dc dynamic.Interface, _ kubernetes.Interface) ([]byte, error) {
+	return mutate(rawPod, ns, injectCWSPodInstrumentation, dc)
+}
+
+func injectCWSPodInstrumentation(pod *corev1.Pod, ns string, _ dynamic.Interface) error {
+	var injected bool
+	defer func() {
+		metrics.MutationAttempts.Inc(metrics.CWSPodInstrumentation, strconv.FormatBool(injected), "")
+	}()
+
+	if pod == nil {
+		metrics.MutationErrors.Inc(metrics.CWSPodInstrumentation, "nil pod", "")
+		return fmt.Errorf("cannot inject CWS instrumentation into nil pod")
+	}
+
+	// sanity check: make sure the provided CWS Injector image name isn't empty
+	if len(config.Datadog.GetString("admission_controller.cws_instrumentation.image_name")) == 0 {
+		metrics.MutationErrors.Inc(metrics.CWSPodInstrumentation, "missing image name", "")
+		// don't touch the pod and leave now
+		return fmt.Errorf("empty CWS Injector image name")
+	}
+
+	// compute init container resources
+	resources, err := initCWSInitContainerResources()
+	if err != nil {
+		metrics.MutationErrors.Inc(metrics.CWSPodInstrumentation, "invalid resources configuration", "")
+		return err
+	}
+
+	// is the namespace targeted by the instrumentation ?
+	if !isNsTargetedByCWSInstrumentation(ns) {
```

Comment:
The check should be done before (almost) anything else, notably `initCWSInitContainerResources()`

Reply:
Updated 👍🏻 I pushed it up for both webhooks, right below the defer.